### ISSUE TITLE
useless_vec: missing spans if vec is not a fn param

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#12101]
### Summary

Clippy does not generate proper diagnostics/spans here which in turn makes rustfix fail...

### Reproducer

I tried this code:

```rust
// run-pass

struct Pair { x: isize, y: isize }

pub fn main() {
    for elt in &(vec![Pair {x: 10, y: 20}, Pair {x: 30, y: 0}]) {
        assert_eq!(elt.x + elt.y, 30);
    }
}
```

I expected to see this happen:
Something pointing inside the vec inside the `&(  .. )`

Instead, this happened:

```
warning: useless use of `vec!`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_vec
  = note: `#[warn(clippy::useless_vec)]` on by default

warning: 1 warning emitted
```
clippy completely fails to get proper spans lol

### Version

```text
rustc 1.77.0-nightly (f688dd684 2024-01-04)
binary: rustc
commit-hash: f688dd684faca5b31b156fac2c6e0ae81fc9bc90
commit-date: 2024-01-04
host: x86_64-unknown-linux-gnu
release: 1.77.0-nightly
LLVM version: 17.0.6
```
